### PR TITLE
RISCV compiler: Set mabi and march via Kconfig options

### PR DESCRIPTION
**Pull request by jaronkelleher**
The mabi and march options of the compiler and linker commands
were previously hardcoded and depended only on the 64BIT config
option. This update allows these flags to be set by the config
options currently available, plus an additional option to
specify the compressed ISA.

Signed-off-by: Jaron Kelleher <jkelleher@fb.com>

### CODE REVIEW

File: cmake/compiler/gcc/target.cmake [hunk 1]
```diff
@@ -49,11 +49,7 @@ elseif("${ARCH}" STREQUAL "arc")
     -mcpu=${GCC_M_CPU}
     )
 elseif("${ARCH}" STREQUAL "riscv")
-  if(CONFIG_64BIT)
-    list(APPEND TOOLCHAIN_C_FLAGS -mabi=lp64 -march=rv64imac -mcmodel=medany)
-  else()
-    list(APPEND TOOLCHAIN_C_FLAGS -mabi=ilp32 -march=rv32ima)
-  endif()
+  include(${ZEPHYR_BASE}/cmake/compiler/gcc/target_riscv.cmake)
```
Comment (SebastianBoe) [inline cmake/compiler/gcc/target.cmake]: include(${CMAKE_CURRENT_LIST_DIR}/target_riscv.cmake)

**Review by SebastianBoe — APPROVED** [review]

File: cmake/compiler/gcc/target.cmake (hunk 1 above)
Comment (jaronkelleher) [reply cmake/compiler/gcc/target.cmake]: I'm AFK until Monday. I'll make this update then

Comment (jaronkelleher) [reply cmake/compiler/gcc/target.cmake]: Complete

**Review by nategraff-sifive — COMMENTED** [review]
Any reason not to just have `CONFIG_RISCV_ARCH` and `CONFIG_RISCV_ABI`? Saves you the logic of building them out of other options and lets the user manually override them with whatever they prefer without thinking through the config options.

**Comment by jaronkelleher** [comment]
> Any reason not to just have `CONFIG_RISCV_ARCH` and `CONFIG_RISCV_ABI`? Saves you the logic of building them out of other options and lets the user manually override them with whatever they prefer without thinking through the config options.

That was another route I had considered going down and am still open to that direction. I chose this approach so that some of that information could be abstracted from the user, and the user can't input an invalid architecture/abi. Though there could be safeguards implemented to prevent that.

I don't have a very strong preference of one way or the other. I can implement my vision of that approach in a different branch and we decide from there?

**Comment by nategraff-sifive** [comment]
I think the ultimate goal should be to get this information out of the Devicetree by default with the ability to manually override it. Starting off with the ISA represented in CMAKE first and foremost as the complete string I think leaves the most flexibility for future evolution of this feature with the least risk of breaking backwards-compatibility in the Kconfig API.

**Comment by carlescufi** [comment]
@nategraff-sifive any objections to merging this?

**Comment by xodus7** [comment]
```CONFIG_FLOAT``` is a good example of a Kconfig option that needs to also modify the -march flags. for example, with #23430 setting one without the other results in a compilation error. Support for builtin atomic operations could be another example in the future.

**Comment by jaronkelleher** [comment]
Any updates here? I'd like to stick with the original idea

**Review by nashif — APPROVED** [review]

**Pull request merged**